File: fmcg-backend/app/models/delivery.py
"""Delivery model. One delivery per order, assigned to a driver."""
from datetime import datetime

from sqlalchemy import DateTime, Enum, ForeignKey, String, Text
from sqlalchemy.orm import Mapped, mapped_column

from app.core.database import Base
from app.models.mixins import TimestampMixin
from app.schemas.common import DeliveryStatus


class Delivery(TimestampMixin, Base):
    __tablename__ = "deliveries"

    id: Mapped[int] = mapped_column(primary_key=True)
    order_id: Mapped[int] = mapped_column(
        ForeignKey("orders.id", ondelete="CASCADE"), unique=True, index=True
    )
    driver_id: Mapped[int] = mapped_column(
        ForeignKey("users.id", ondelete="RESTRICT"), index=True
    )
    status: Mapped[DeliveryStatus] = mapped_column(
        Enum(DeliveryStatus, name="delivery_status"),
        default=DeliveryStatus.PENDING,
        index=True,
    )
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    proof_photo_url: Mapped[str | None] = mapped_column(String(500), nullable=True)
    assigned_at: Mapped[datetime] = mapped_column(DateTime(timezone=True))
    delivered_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )

File: fmcg-android/app/src/main/java/com/fmcg/app/domain/model/Delivery.kt
package com.fmcg.app.domain.model

enum class DeliveryStatus(val api: String, val label: String) {
    PENDING("pending", "Pending"),
    OUT_FOR_DELIVERY("out_for_delivery", "Out for delivery"),
    DELIVERED("delivered", "Delivered"),
    FAILED("failed", "Failed"),
    UNKNOWN("unknown", "Unknown");

    companion object {
        fun fromApi(value: String): DeliveryStatus =
            entries.firstOrNull { it.api == value } ?: UNKNOWN
    }
}

data class Delivery(
    val id: Int,
    val orderId: Int,
    val driverId: Int,
    val status: DeliveryStatus,
    val notes: String?,
    val proofPhotoUrl: String?,
    val assignedAt: String,
    val deliveredAt: String?,
)
